src/NSwag.AspNetCore/Middlewares/SwaggerDocumentMiddleware.cs

```diff
@@ -76,29 +79,51 @@ protected virtual async Task<string> GetDocumentAsync(HttpContext context)
             }
 
             var apiDescriptionGroups = _apiDescriptionGroupCollectionProvider.ApiDescriptionGroups;
-            if (apiDescriptionGroups.Version == Volatile.Read(ref _version) && _swaggerJson != null)
+
+            // The bellow code uses the swagger document Post Processing logic to get an actual
+            // host and basePath values. It looks like it's better to use a more lightweight DTO in PostProcess callback.
+            var doc = new SwaggerDocument();
+            _settings.PostProcess?.Invoke(doc, context.Request);
+            var key = new SwaggerDocumentKey(doc.BasePath ?? context.Request.PathBase, doc.Host ?? context.Request.Host.Host);
```

Comment:
Can't we just cache with host and port to avoid calling PostProcess? From what is the base path derived from? What are the scenarios?

Reply:
In one of my demo environments, we expose our API on a subpath (e.g. https://mydomain.com/my-demo-application/), so I have to rely on the subpath value.

I completely agree the initial changes look like a huge hack and we have to rethink it towards the more appropriate solution. Speaking of the machinery behind the code, let me put some words here:

When a user performs a request, the middleware doesn't know which version of a JSON document it should return. So, it should look up the right JSON document from the dictionary taking into account the source of the HTTP request(internal network access or it's a request behind the reverse proxy). To allow the middleware to make the right lookup we provide a callback. In the current PR, I rely on the logic of the PostProcess callback.

So, the code of the callback I use looks like:
```
config.PostProcess = (document, request) =>
					{
						document.Host = ExtractHost(request) + ":" + ExtractPort(request);
						document.BasePath = ExtractBasePath(request);
						document.Schemes.Clear();
						SwaggerSchema httpScheme = ExtractProto(request).EndsWith("s", StringComparison.OrdinalIgnoreCase)
							                           ? SwaggerSchema.Https
							                           : SwaggerSchema.Http;
						document.Schemes.Add(httpScheme);
					};
```
ExtractHost/ExtractPort/ExtractProto implementations are straightforward and look like
```
private string ExtractProto(HttpRequest request) => FirstOrDefault(request.Headers["X-Forwarded-Proto"]) ?? request.Scheme;
```

Perhaps, to avoid passing an empty document we just need to add another callback that allows getting rid of calling _PostProcess_ to build the lookup key.
e.g.
instead of
```
var doc = new SwaggerDocument();
_settings.PostProcess?.Invoke(doc, context.Request);
var key = new SwaggerDocumentKey(doc.BasePath ?? context.Request.PathBase, doc.Host...
```
use something like
`Func<(string basePath, string host, int port)> Request;`
So the user code looks like
```
.UseSwagger(config => {
    config.Request  = () => (ExtractBasePath(request), ExtractHost(request), ExtractPort(request));
                        // or
    config.Request  = () => /* build an url that indicates the application locatiion, e.g. https://mydomain.com/my-demo-application/ and use it instead of tuple */;
});
```

Reply:
Maybe something like `Func<HttpRequest, string> CreateDocumentCacheKey { get; set; }` on the settings class with a default implementation which just returns `string.Empty` (always the same key = current behavior) - this way you can customize it completely...

Reply:
LGTM :)